### PR TITLE
[WIP] add buffered chunked streaming for nemo force aligner

### PR DESCRIPTION
**Pull request by Slyne**
# What does this PR do ?
To support buffered chunked streaming inference for nemo neural force aligner.

**Collection**: [Note which collection this PR will affect]
ASR

# Changelog 
- add an argument (keep_logits default is False) for transcribe function in `FrameBatchASR` to keep the inference logits.
- return the logits(log probs) if keep_logits is set to True.

# Usage

```python
python align.py \
        pretrained_name="stt_en_citrinet_1024_gamma_0_25" \
        batch_size=32 \
        model_downsample_factor=8 \
        manifest_filepath=$manifest \
        use_buffered_chunked_streaming=true \
        output_dir=./buffered_streaming_test_pred \
        total_buffer_in_secs=10 \
        align_using_pred_text=true
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tools/nemo_forced_aligner/align.py [hunk 1]
```diff
@@ -194,6 +209,38 @@ def main(cfg: AlignmentConfig):
...
+        feature_stride = model_cfg.preprocessor['window_stride']
+        model_stride_in_secs = feature_stride * cfg.model_downsample_factor
+        total_buffer = cfg.total_buffer_in_secs
+        chunk_len = float(cfg.chunk_len_in_secs)
+        tokens_per_chunk = math.ceil(chunk_len / model_stride_in_secs)
+        mid_delay = math.ceil((chunk_len + (total_buffer - chunk_len) / 2) / model_stride_in_secs)
+        logging.info(f"tokens_per_chunk is {tokens_per_chunk}, mid_delay is {mid_delay}")
+
+        model = FrameBatchASR(
+            asr_model=model, frame_len=chunk_len, total_buffer=cfg.total_buffer_in_secs, batch_size=cfg.batch_size,
```
Comment (erastorgueva-nv) [inline tools/nemo_forced_aligner/align.py:237]: Please make a separate parameter, maybe called `chunk_batch_size`, and default value maybe 32 to pass in as the `FrameBatchASR` `batch_size` parameter. This parameter corresponds to the number of buffers in a batch (with the buffers coming from the same audio file).

**Review by erastorgueva-nv — CHANGES_REQUESTED** [review]
Thank you for this PR Slyne. Please modify the FrameBatchASR parameter as described in the comment.

**Comment by erastorgueva-nv** [comment]
cc @jbalam-nv for visibility on changes to streaming_utils. I don't think they will break anything.
We need to add self.decoder, self.cfg and self.preprocessor attributes to FrameBatchASR so we can use them downstream in NFA. We also add some keep_logits flags but they are False by default.

**Review by jbalam-nv — DISMISSED** [review]
LGTM

**Review by erastorgueva-nv — APPROVED** [review]

**Review by erastorgueva-nv — APPROVED** [review]
LGTM, this is the same as before

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 2]
```diff
@@ -772,19 +775,32 @@ def _get_batch_preds(self):
...
         for pred in self.all_preds:
             decoded = pred.tolist()
             self.unmerged += decoded[len(decoded) - 1 - delay : len(decoded) - 1 - delay + tokens_per_chunk]
-        return self.greedy_merge(self.unmerged)
+        hypothesis = self.greedy_merge(self.unmerged)
+        if not keep_logits:
+            return hypothesis
+
+        all_logits = []
+        for log_prob in self.all_logits:
```
Comment (titu1994) [inline nemo/collections/asr/parts/utils/streaming_utils.py:798]: Could you put this in a function? I feel like it gets repeated a lot

**Review by titu1994 — COMMENTED** [review]
Overall it looks awesome ! Minor comments, feel free to ignore for now
